### ISSUE TITLE
Question on mask dimension for the code in appendix C

### ISSUE THREAD

**Issue by xyang2013** [rasbt/reasoning-from-scratch#67]
```python
queries = self.W_query(x)  # (b, num_tokens, num_heads * head_dim)
queries = queries.view(b, num_tokens, self.num_heads, self.head_dim).transpose(1, 2)

# dim of queries: (b, self.num_heads, num_tokens, self.head_dim)

keys = self.W_key(x)       # (b, num_tokens, num_kv_groups * head_dim)
keys_new = keys.view(b, num_tokens, self.num_kv_groups, self.head_dim).transpose(1, 2)
keys = torch.cat([prev_k, keys_new], dim=2)
keys = keys.repeat_interleave(self.group_size, dim=1)

# dim of keys:  (b, self.num_heads, (prev_k_number_tokens + num_tokens), self.head_dim)

attn_scores = queries @ keys.transpose(2, 3)

# dim of attn_scores: (b, self.num_heads, num_tokens, (prev_k_number_tokens + num_tokens))
```

Then for the mask:

```python
# Shape (1, 1, num_tokens, num_tokens) to broadcast across batch and heads
mask = mask[None, None, :, :]
```

The dimensions do not seem to match. I am not sure how the mask is applied here.

**Comment by rasbt** [comment]
Hi there, and thanks for the question. When I understand correctly, the issue in specifically in the KV cache path, i.e., where we have a mask shape `(1, 1, num_tokens, num_tokens)`?

The code comment there might be a bit misleading. For the non-cached part it is always correct. But for the cached part, it is only correct for the prefill stage. Then, in each subsequent forward (generation) step, we have a mask size of `(1, 1, 1, total_context_length)`.


So, let's say we start with 2 input tokens (A and B), we have

Computed keys [ A  B ]

And the mask

```    
     A  B
A  [ 1  0 ]
B  [ 1  1 ]
```

for prefill. Then, in the first generation step, we have:

Cached keys [ A  B ]
Computed keys [ C ]
All keys: [ A  B | C ]

Mask: 

```
C  [ 1  1 | 1 ]
```

and so on.


Does this help clarify? Perhaps I should add more explanation to the code, e.g., changing


```
# Shape (1, 1, num_tokens, num_tokens) to broadcast across batch and heads
mask = mask[None, None, :, :]  # broadcast mask
```

to

```
# Prefill (no cache): mask starts as (num_tokens, num_tokens)
# Cached decoding: mask starts as (num_tokens, prev_k_number_tokens + num_tokens)
#
# We add two leading dimensions so the mask becomes
# (1, 1, num_tokens, num_tokens) during prefill and
# (1, 1, num_tokens, total_key_tokens) during cached decoding.
# These extra dimensions let PyTorch broadcast the same mask
# across all batches and attention heads when applying it to
# attn_scores of shape (batch, num_heads, num_tokens, total_key_tokens).
mask = mask[None, None, :, :]
```

**Comment by xyang2013** [comment]
Thanks, @rasbt. That makes sense now. During prefill, the mask has a shape of (1, 1, num_tokens, num_tokens), and for subsequent decoding steps, it becomes (1, 1, 1, total_context_length).

By the way, it seems that `next_cache` in `Qwen3Model(nn.Module)` isn’t actually needed.

**Comment by rasbt** [comment]
Glad this helped! Btw regarding your second question:

> By the way, it seems that next_cache in Qwen3Model(nn.Module) isn’t actually needed.

Do you mean in `GroupedQueryAttention` when we are ***not*** using KV caching, the `next_cache = (keys, values)` becomes redundant? In

```python
class GroupedQueryAttention(nn.Module):
# ...
        if cache is not None:
            prev_k, prev_v = cache
            keys = torch.cat([prev_k, keys_new], dim=2)
            values = torch.cat([prev_v, values_new], dim=2)
        else:
            start_pos = 0  # reset RoPE
            keys, values = keys_new, values_new
        next_cache = (keys, values)
```

I did this for simplicity, but yes, I suppose we could change it to 


```python
class GroupedQueryAttention(nn.Module):
# ...
        if cache is not None:
            prev_k, prev_v = cache
            keys = torch.cat([prev_k, keys_new], dim=2)
            values = torch.cat([prev_v, values_new], dim=2)
            next_cache = (keys, values)
        else:
            keys, values = keys_new, values_new
            next_cache = None
```

But I'd say the current version is a bit easier on `torch.compile` perhaps.

**Comment by xyang2013** [comment]
Thanks @rasbt . Sorry, I was referring to your previous version of the code:
```
        next_cache = []
        for i, block in enumerate(self.trf_blocks):
            blk_cache = cache.get(i) if cache else None
            x, new_blk_cache = block(x, mask, self.cos, self.sin,
                                     start_pos=pos_start,
                                     cache=blk_cache)
            if cache is not None:
                cache.update(i, new_blk_cache)
            next_cache.append(new_blk_cache)

        x = self.final_norm(x)
        logits = self.out_head(x.to(self.cfg["dtype"]))
        return logits
```

I just checked - the new version does have next_cache.

**Comment by rasbt** [comment]
Thanks for clarifying.

> I just checked - the new version does have `next_cache`.

I think you mean "does not have", right? And if yes, then you are right. The new version should not have it. (The publisher added chapter 3 on the website yesterday, and the updated `Qwen3Model` code in the book should not have the redundant `next_cache`.)

Anyways, thanks for the good discussion. That was pretty useful feedback, and I will update the code comment later as suggested.
